### DISCUSSION TITLE
単体テストの考え方/使い方 4.5〜5.1

### DISCUSSION THREAD

**Discussion by dameleon** [atman-inc/training#98] [General]
- 4.5: ソフトウェア・テストにおけるよく知られた概念
- 5.1: モックとスタブの違い

行けるとこまで。はみ出たら次週。

上記辺りののディスカッション。
不明点、相談点、確認点、勉強になったポイントなどは各自ディスカッションのコメントをお願いします。

**Comment by yutafukuchi-1015** [comment]
## 4.5.1
#### 単体テスト
- テストケースが多い
  - そのため迅速なフィードバックが重要

#### E2Eテスト
- テストケースが少なくコード量が多い
  - 退行に対する保護が重要

## 4.5.2
#### ブラックボックステスト
ユーザー目線で何を行いたいかを基準にテストが作成されている(?)
基本的にはブラックを使用する

#### ホワイトボックステスト
コードの細部までを見るテスト
ブラックより詳細なテストが必要なコードの場合に、こちらを使用する

**Comment by xxManatoxx** [comment]
## 4.5 ソフトウェアテストにおけるよく知られた概念

### 4.5.1 テスト・ピラミッド

- 3種類のテストで構成されたピラミッドの図を用いて表現されることが多い
- 3種類のテスト
  - 単体テスト
  - 統合テスト
  - E2Eテスト
- ピラミッドで表現していること
  - 横幅
    - テストケースの数
  - 縦幅
    - エンドユーザへの近さ
- 層によって異なること
  - 退行に対する保護と迅速なフィードバックとのバランス感
    - 単体テスト
      - 迅速なフィードバックが重要
    - 統合テスト
      - 単体テスト、E2Eテストの中間で退行に対する保護と迅速なフィードバックのどちらも重要になる
    - E2Eテスト
      - 退行に対する保護が重要
  - リファクタリングへの耐性
    - どの層でも必ず備えるべき
- テストピラミッドの例外
  - ビジネスルールや複雑なことを扱わない時
    - 各テストで用意されるテストケースがほぼ同じになる
  - 1つのプロセス外依存（例ではDB）しか扱わない時
    - E2Eテストがより多く用意したほうが効果が高い事がある

### 4.5.2 ブラック・ボックス・テストとホワイト・ボックス・テスト

- ブラック・ボックス・テスト
  - システム内部の処理を考慮せず、システムの仕様や要求を元に作成されるテスト
  - 長所
    - 要求・仕様から作成するためより振る舞いに注視したテストケースになる
  - 短所
    - 要求・仕様から見つけられないエラーが出てくることがある
- ホワイト・ボックス・テスト
  - システムの内部をテストするため、ソースコードから作成される（内部の処理を理解した上でその処理が本当に正しく行われているか等）
  - 長所
    - 実装から作成するため、ブラック・ボックス・テストで見逃しそうなエラーを検知出来る
  - 短所
    - ソースコードの分析が必要
    - テスト対象の内部的なコードと深く結びつき、すぐに壊れるようになる
    - ビジネスにおける意味のある振る舞いを認識しづらくなる傾向がある

## 5章 モックの利用とテストの壊れやすさ

- この章では単体テストにおけるロンドン学派と古典学派の考え方の違いについて主に見ていくことになる

### 5.1 モックとスタブの違い

#### 5.1.1 テストダブルの種類

- テストダブル
  -  テストでしか使われない偽りの依存として表現されるすべてのものを包括的に意味するもの
  - 語源は映画のスタントマン（Stunt double）
  - 使用される理由
    - テストを円滑に実施できるようにするため
  - 種類
    - ダミー
    - スタブ
    - スパイ
    - モック
    - フェイク
  - 大きく分類すると
    - モック
      - モック、スパイ
    - スタブ
      - スタブ、ダミー、フェイク
- モックとスタブの違い
  - モック
    - テスト対象システムから、その依存に向かって行われる「外部に向かう」コミュニケーション（出力）を模倣
    - テスト対象システムとその依存とのコミュニケーションを模倣するだけでなく、検証も行う
    - 例
      - メールの送信
  - スタブ 
    - テスト対象システムに向かって行われる「内部に向かう」コミュニケーション（入力）を模倣
    - テスト対象システムとその依存とのコミュニケーションを模倣のみ行う
    - 例
      - データベースからデータを取得する
- テストダブルの種類の違いは細かな実装部分に関するもので、そこまで重要性は無い
  - 例えばスパイのことを手書きのモックと呼ぶことがある（モックはモックフレームワークを使用、スパイは開発者自身が実装するため）
  - スタブ、ダミー、フェイクはどのくらい知的に振る舞えるか
  - ダミーはハードコーディングされたダミーの値の事（null、文字列など）
  - スタブは設定によって返す結果を変える完全に自立した依存として振る舞う
  - フェイクは基本的にスタブと同じだが、フェイクはまだ存在しないものを置き換えるという動機の部分が異なる

**Comment by y-kaneda** [comment]
## 4.5 ソフトウェア・テストにおけるよく知られた概念

### 4.5.1 テスト・ピラミッド
- テストの種類およびそれぞれのケース量の割合の概念
  - 単体・統合・E2E は下に行けば量が多く、上へ行けばエンドユーザの観点に近い
- テストの層によって退行に対する保護・リファクタリングへの耐性・迅速なフィードバックの3つのバランスの優先度が変わってくる
  - 高い層(ユーザ寄り)は退行に対する保護、低い層(単体テスト)は迅速なフィードバックがそれぞれ求められる
  - リファクタリングへの耐性は共通して求められる
- E2Eテストの存在意義は単体テスト・統合テストで担保できない、重要でバグを発生させたくない機能のために行うもの
- テスト対象の機能が複雑なドメイン等を扱っていない場合には各層のテストの量のバランスは変わってくる

### 4.5.2 ブラック・ボックス・テストとホワイト・ボックス・テスト
- ブラック・ボックス・テスト
  - システム内部構造を意識せずに行うテスト
  - テスト対象を how ではなく what の観点で見る
- ホワイト・ボックス・テスト
  - システムの内部を検証するためのテスト
  - 仕様や要求からではなくソースコードから作成される
  - 仕様に基づくブラック・ボックス・テストでは見落としてしまうようなエラーが検出できる
  - プロダクションコードと深く結びつくため、テストケースが壊れやすく、リファクタリングへの耐性が低い
- まずはブラック・ボックス・テストを選択する
- テストケースからビジネス要求が分からない場合、壊れやすいホワイト・ボックス・テストになっている
- ホワイト・ボックス・テストの観点で分析し、気付けていなかった分岐等を見つけてブラック・ボックス・テストのケースに活かすのはアリ

# 第5章 モックの利用とテストの壊れやすさ
- モックは優れたツールである反面、テストを壊れやすくする傾向がある

## 5.1 モックとスタブの違い
- テスト・ダブル
  - モック
    - (スパイ)
  - スタブ
    - (ダミー)
    - (フェイク)

### 5.1.1 テストダブルの種類
- テストダブル
  - テストでしか使われない偽りの依存全て
  - スタントマン(stunt double)からきている
- モック
  - テスト対象システムからその依存に向かって行われる、外部に向かうコミュニケーションを模倣
  - スパイはフレームワークを用いず開発者が実装する
  - テスト対象システムとその依存のコミュニケーションを模倣だけではなく検証も行う
    - 例えばきちんとメール送信のライブラリの関数が呼ばれてるよね、等
- スタブ
  - 依存からテスト対象システムに向かって行われる、内部に向かうコミュニケーションを模倣
  - ダミーはnullやリテラル等、一時しのぎのシンプルなハードコーディング
  - スタブは設定によりシナリオごとに動的に返す値を変えられる
  - フェイクはスタブに近いが、まだ存在しない依存を置き換えるために作成される
  - テスト対象システムとその依存のコミュニケーションを模倣だけする

### 5.1.2 複数の意味を持つ「モック」
- テスト・ダブル全体が「モック」と呼ばれがち
- モックのライブラリから提供されるモックのクラスのことだけを意味していることがある

### 5.1.3 スタブとのコミュニケーションは検証しない
- スタブの役割はテストのために必要なデータを提供するだけ
- ダミーデータの生成等を検証することはテストの目的に関係ない
  - 過剰検証(over-specification)

### 5.1.4 モックとスタブの両方の性質を持ったテスト・ダブル
- あるケースで呼び出しも結果も検証が必要な場合

### 5.1.5 モックとスタブがそれぞれどのようにコマンドとクエリに結びつくのか？
- コマンド・クエリ分離(Command Query Separation: CQS)の原則
  - コマンド
    - 戻り値のない、副作用を起こすメソッド
    - モックに紐づく
  - クエリ
    - 副作用は起こさず、何らかの値を返すメソッド
    - スタブに紐づく

**Comment by nmurakami0** [comment]
4.5

- 4.5.1: テスト・ピラミッド
  - ピラミッドの層が異なれば、regressionに対する保護と迅速なフィードバックとのバランスにおいてどうするのが最善かが変わる
    - 上の層ほどregressionに対する保護を重視すべき
    - 下の層ほど迅速なフィードバックを重視すべき
- 4.5.2
  - ホワイトボックステストよりブラックボックステストを重視すべき
    - regressionへの体制は、その声質を備えるか否かのどちらかしか選択できないため

5.1

- モックとスタブの違い
  - モックはテスト・ダブルのこと
    - テスト・ダブル: プロダクションコードに含まれず、テストでしか使われない偽りの依存として表現されるすべてのものを包括的に意味する
    - 5種類あり、2つに大別出来る: 
      - モック系: モック、スパイ
        - テスト対象システムからその依存に向かって行われる外部に向かうコミュニケーションを模倣し、その検証に用いられる
        - e.g., メールの送信部分をモックにする
        - モックはモック・フレームワークの助けを借りて生成されるが、スパイは開発者の手で実装される
      - スタブ系: ダミー、スタブ、フェイク
        - 依存からテスト対象システムに向かって行われる内部に向かうコミュニケーションの模倣に使われる
        - e.g., データベースからのデータの取得部分をスタブにする
        - ダミーはシンプルなハードコーディングされたもの。

内部か外部かと、検証を行うかどうかが大きな違い。

上記踏まえると、Web Mockは命名が正しそう
( https://github.com/bblimke/webmock )

HTTPリクエスト(GET)経由のデータの取得は、上の定義だと基本的にはスタブになる？

rspecのdoubleやspyは上の定義と一致しているか？
(確かspyは指定した一部のメソッドだけ模倣し、それ以外は本来の実装を用いるはず)

**Comment by myamamoto111** [comment]
## 4.5 ソフトウェア・テストにおけるよく知られた概念

### 4.5.1 テストピラミッド
- 単体テスト
- 統合テスト
- E2Eテスト

テストケースの数とエンド・ユーザの観点の近さで表すピラミッド層

### 4.5.2 ブラック・ボックス・テストとホワイト・ボックス・テスト
- ブラック・ボックス・テスト
  - システムの内部構造を知ることなしに検証するソフトウェア・テストの手法
  - 退行に対する保護：劣
  - リファクタリングへの耐性：優
- ホワイト・ボックス・テスト
  - アプリケーションが内部的に行なっていることを検証する
  - 退行に対する保護：優
  - リファクタリングへの耐性：劣

テストケースを作成する場合、ブラック・ボックス・テスト
問題領域にとって意味のある振る舞いのみを検証の対象とする

テストケースを分析する場合、ホワイト・ボックス・テスト

## 第５章　モックの利用とテストの壊れやすさ
- **モックとスタブの違い**
- **観察可能な振る舞いと実装の詳細の定義**
- **モックの利用とテストの壊れやすさの関係**
- **リファクタリングへの耐性を損なわずにモックを使う方法**

## 5.1 モックとスタブの違い
### 5.1.1 テストダブルの種類
- テストダブルとは、テストでしか使われない偽りの依存として表現される全て包括したもの
  - ダミー
  - スタブ
  - スパイ
  - モック
  - フェイク
実際には、モックとスタブの２つに分けることが出来る
- モック
  - 外部に向かう出力の模倣
- スタブ
  - 内部に向かう出力の模倣

### 5.1.2 複数の意味を持つ「モック」
- 道具としてのモック
- テストダブルとしてのモック

### 5.1.3 スタブとのコミュニケーションは検証しない
- 過剰検証

### 5.1.4 モックとスタブの両方の性質を持ったテスト・ダブル
- モックと呼ばれる

### 5.1.5 モックとスタブがそれぞれどのようにコマンドとクエリに結びつくのか？
コマンド・クエリ分離（Command Query Separation: CQS）の原則
- コマンド
  - 副作用を起こす
  - 戻り値なし
- クエリ
  - 副作用なし
  - 戻り値を返す

- コマンドの代わりとして使われるのがモック
- クエリの代わりとして使われるのがスタブ

モック、ダミーくらいしか意識したことなかったけど、色々と定義があるんだな。

**Comment by koushisa** [comment]
## 4

### ホワイトボックステスト
- あんまり書きたくはない
- ブラックボックスでカバーしたい
- opinion
  - 必要なら、FSM(ステートマシン)化してホワイトボックステストはFSMから自動生成したい
  - 組み込みシステムなど生存期間が長く改修が難しいケースでは自動生成が役にたつイメージ
  - 事例
    - ヘイ株式会社のSTORESの予約システムフロントエンドは状態遷移が複雑なのでXStateでホワイトボックステストしてるっぽい
      - [XStateを支える概念と実装方法について](https://product.st.inc/entry/2022/04/14/105719)
    - awsのダッシュボード系とかamplifyの認証周りの内部実装でも確かXStateでホワイトボックステストしてる

**Comment by nikuteresa** [comment]
## 4.5
一般的に知られているソフトウェア・テストの概念
- テスト・ピラミッド
- ブラック・ボックス・テスト
- ホワイト・ボックス・テスト

テスト・ピラミッド
- 大きく３つのテスト要素に分類されている
- 下に位置する要素ほどテストケースの数が多い
- 上に位置する要素ほどエンドユーザの視点に近い検証を行う
- どのテスト要素もリファクタリングへの耐性を備える必要がある
- 上に位置するテスト要素ほどテストケースが少ないのはテストにかかるコストが多いのが理由の１つ

ブラック・ボックス・テスト
- 実装はみないで設計書からテストを作成する
- 行われたことが正しいかを検証
- テストケースが足りない状況に陥りやすい(退行に対する耐性が低い)

ホワイト・ボックス・テスト
- 実装を見てテストを作成する
- 内部的な処理自体を検証をする
- リファクタリング耐性が低い
- false positive が発生しやすい
- テスト対象の正常な振る舞いを認識しづらい

できるだけブラック・ボックス・テストを作るべき

## 5.1
スタブ
- テスト・ダブルに含まれているうちの１つ
- 依存をテスト対象に向かって肩代わりする
- 例）テストがデータの取得を行うときに取得される値をあらかじめ設定しておく（その取得されたデータの値によって処理が分岐する場合などに便利）


モック
- テスト対象からの依存を肩代わりする
- 例）テストがメールを送る場合にメールを送ったことにする（実際に送らなくてもエラーにならないで済むなどのメリット）

疑問点
- メールを送り、メールの送信結果として true false を返すという依存に対してのテスト・ダブルはモックなのかスタブなのか
	- 全く同じ内容のテスト・ダブルだったしてもテスト観点によって呼び方が変わる？
